src/ziggurat/init.clj

```diff
@@ -28,38 +27,101 @@
+  (messaging-consumer/start-subscribers (get args :stream-routes)))
+
+(defn- start-rabbitmq-producers [args]
+  (start-rabbitmq-connection args)
+  (messaging-producer/make-queues (get args :stream-routes)))
+
+(defn start-stream [args]
+  (start-rabbitmq-producers args)
+  (start* #{#'streams/stream} args))
+
+(defn start-management-apis [args]
+  (start-rabbitmq-connection args)
+  (start* #{#'server/server} (dissoc args :actor-routes)))
+
+(defn start-server [args]
+  (start-rabbitmq-connection args)
+  (start* #{#'server/server} args))
+
+(defn start-workers [args]
+  (start-rabbitmq-producers args)
+  (start-rabbitmq-consumers args))
+
+(defn- stop-rabbitmq-connection []
+  (mount/stop #'messaging-connection/connection))
+
+(defn stop-workers []
+  (stop-rabbitmq-connection))
+
+(defn stop-server []
+  (mount/stop #'server/server)
+  (stop-rabbitmq-connection))
+
+(defn stop-stream []
+  (mount/stop #'streams/stream)
+  (stop-rabbitmq-connection))
+
+(defn stop-management-apis []
+  (mount/stop #'server/server)
+  (stop-rabbitmq-connection))
+
+(def valid-modes-fns
+  {"api-server"     [start-server stop-server]
+   "stream-worker"  [start-stream stop-stream]
+   "worker"         [start-workers stop-workers]
+   "management-api" [start-management-apis stop-management-apis]})
+
+(defn- execute-function
+  ([modes fnk]
+   (execute-function modes fnk nil))
+  ([modes fnk args]
+   (doseq [mode (-> modes
+                    (or (keys valid-modes-fns))
+                    sort)]
+     (if (nil? args)
+       ((fnk (get valid-modes-fns mode)))
+       ((fnk (get valid-modes-fns mode)) args)))))
+
+(defn start-common-states []
   (start* #{#'config/config
             #'statsd-reporter
-            #'sentry-reporter})
-  (actor-start-fn)
-  (start* #{#'messaging-connection/connection} {:stream-routes stream-routes})
-  (messaging-producer/make-queues stream-routes)
-  (messaging-consumer/start-subscribers stream-routes)      ;; We want subscribers to start after creating queues on RabbitMQ.
-  (start* #{#'server/server
-            #'nrepl-server/server
-            #'streams/stream}
-          {:stream-routes stream-routes
-           :actor-routes  actor-routes}))
+            #'sentry-reporter
+            #'nrepl-server/server}))
 
-(defn stop
-  "Calls the Ziggurat's state stop fns and then actor-stop-fn."
-  [actor-stop-fn]
+(defn stop-common-states []
   (mount/stop #'config/config
               #'statsd-reporter
               #'messaging-connection/connection
-              #'server/server
-              #'nrepl-server/server
-              #'streams/stream)
-  (actor-stop-fn)
-  (mount/stop #'config/config))
+              #'nrepl-server/server))
+
+(defn start
+  "Starts up Ziggurat's config, reporters, actor fn, rabbitmq connection and then streams, server etc"
+  [actor-start-fn stream-routes actor-routes modes]
+  (start-common-states)
+  (actor-start-fn)
+  (execute-function modes
+                    first
```

Comment:
why `first` or `second`? using a map like `{:start-fn x :stop-fn y}` make it clearer. wdyt?

Reply:
Make sense. Will change it.